### PR TITLE
[backend/frontend] Be able to add images in overview of some entity types (#issue/2540)

### LINKED ISSUE #2540: Be able to add images in overview of some entity types
This feature should be designed as generic in the backend (any types of entity can have images) but in the frontend (overview / cards), only those first types of entities are concerned:

- Threat Actor Individual
- Threat Actor Group
- Intrusion Set
- Malware
- Tool
- Individual
- Organization

**In Data tab :**

- [ ] **`MVP`** Adding a “Pictures” section
    - each uploaded picture (through the “uploaded files” section) is represented by a line in the “Pictures management” section with:
        - a miniature of the Picture
        - the description
        - order
    - the “in Caroussel” checkbox
    - a burger button to manage or delete the picture
    - At the top of the “Pictures management” section are 2 Combo boxes:
        - One to choose which picture will be displayed in the Card view
        - `Nice-to-have` One to choose the caroussel mode : Static=the caroussel does not rotate automatically, Dynamic= the caroussel rotates automatically
    - By default, the list is sorted by Order
- [ ] **`MVP`**Managing a Picture:
    - when clicking on the burger menu of a Picture, it opens an “Manage the Picture” form:
        - field description
        - field Order (with an Information : “Order for the caroussel display)
        - a toggle “in Caroussel” to define if the Picture is in the caroussel or not

!https://share.balsamiq.com/c/cBxJq4zLwKm42bizBjYFin.png

**In Overview:**

- [ ] **`MVP`** The caroussel contain all picture with the “in Caroussel” toggled on.
- [ ] **`MVP`** The caroussel default position is set to display the Picture with the lowest Order value
- [ ] **`MVP`** When hovering a picture in the caroussel, you get the description in a tooltip
- [ ] `Nice-to-have` If no picture has been added, a placeholder is displayed in the caroussel
- [ ] `Nice-to-have` With the caroussel, 3 buttons:
    - remove the displayed picture from caroussel (toggle off the “in Overview”)
    - modify the displayed picture: open the edit form for the picture
    - add a picture to the Overview

!https://share.balsamiq.com/c/bmdgTsZ2yJY8HL754gaPSq.png

**In Card view:**

- [ ] **`MVP`** Picture chosen to be “Picture for Card” in the “Data/Picture management” section is displayed in the card
    - ⚠️ It must be able to handle portait or landscapes pictures.

!https://share.balsamiq.com/c/2vC5sq1t8fEv9jx6ghh1pU.png

### PR DESCRIPTION
**Pull request by SarahBocognano**
This issue is the following of Threat Actor Individual Image Carousel. Now we can manage the pictures uploaded in the Data Section, select their order in the carousel, add a description and choose if we want it to be in the carousel or not.

### Proposed changes

* Add a picture management section to be able to order, add description and select carousel selection or not

### Related issues

* https://github.com/OpenCTI-Platform/opencti/issues/2540

### Checklist

- [ ] I consider the submitted work as finished
- [ ] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yassine-ouaamou** [comment]
Here are some feedbacks we can talk about tomorrow:

- Last History is unreadable
image

- No limit in Image size (maybe store thumbnails or smaller version and show it because loading a 15Mo picture takes time)
- Alignment of cells in picture management
image

- "In Carousel" checkbox becomes false after editing Description or Order

File: opencti-platform/opencti-dev/Backend start.run.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Backend start" type="js.build_tools.npm">
-    <package-json value="$PROJECT_DIR$/opencti-platform/opencti-graphql/package.json" />
+    <package-json value="$PROJECT_DIR$/opencti/opencti-platform/opencti-graphql/package.json" />
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-dev/Backend start.run.xml]: why this change ?

File: opencti-platform/opencti-front/src/private/components/common/files/PictureManagementEdition.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        </IconButton>
+        <Typography variant="h6" classes={{ root: classes.title }}>
+          {t('Update a picture')}
+        </Typography>
+        <div className="clearfix" />
+      </div>
+      <div className={classes.container}>
+        <Formik
+          initialValues={{
+            id: picture.id,
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/files/PictureManagementEdition.tsx:125]: can the picture id change ? otherwise I don't think it should be in the form, we could directly use `picture.id`

File: opencti-platform/opencti-front/src/private/components/common/files/PictureManagementViewer.tsx [hunk 3]
```diff
@@ -52,8 +74,18 @@ const PictureManagementViewer: FunctionComponent<PictureManagementViewerProps> =
         </Typography>
         <div className="clearfix" />
         <Paper classes={{ root: classes.paper }} variant="outlined">
-          {data.images && data.images.length > 0 ? (
-            <List></List>
+          {images && images.length > 0 ? (
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/files/PictureManagementViewer.tsx:85]: images is defined (`data.images ?? []`) so we can directly test `images.length > 0`

File: opencti-platform/opencti-front/src/private/components/threats/threat_actors_individual/ThreatActorIndividualDetails.tsx [hunk 4]
```diff
@@ -81,6 +82,9 @@ const ThreatActorIndividualDetails: FunctionComponent<ThreatActorIndividualDetai
     threatActorIndividualData,
   );
 
+  const image = data.images ?? [];
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/threats/threat_actors_individual/ThreatActorIndividualDetails.tsx]: maybe rename this variable `images` since it's an array ?

File: opencti-platform/opencti-graphql/src/resolvers/stixDomainObject.js [hunk 5]
```diff
@@ -96,7 +96,7 @@ const stixDomainObjectResolvers = {
           () => pubSubAsyncIterator([bus.EDIT_TOPIC, bus.CONTEXT_TOPIC]),
           (payload) => {
             if (!payload) return false; // When disconnect, an empty payload is dispatched.
-            return payload.user.id !== context.user.id && payload.instance.id === id;
+            return payload.instance.id === id;
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/resolvers/stixDomainObject.js:101]: why this change ?

**Comment by Archidoit** [comment]
When we click in the overview of a threat actor individual, the image displayed is the first one, but the points at the bottom don't correspond (the second point is highlighted in the capture below whereas the image is the first one): 
image


@SarahBocognano Yes I know, it's how the carousel library choosed to display the selected image. It's weird to not highlight it, but they choose to shadow the selected one instead

**Comment by Archidoit** [comment]
Missing translations: 'Pictures management', 'in carousel' and 'update a picture'.
image

image

File: opencti-worker/src/config.yml.sample
Comment (SouadHadjiat) [inline opencti-worker/src/config.yml.sample:1]: why delete it ?

File: opencti-platform/opencti-graphql/src/domain/stixCoreObject.js [hunk 6]
```diff
@@ -436,6 +436,7 @@ export const stixCoreObjectImportDelete = async (context, user, fileId) => {
       prevent_indexing: true,
       context_data: contextData
     });
+    await notify(BUS_TOPICS[ABSTRACT_STIX_DOMAIN_OBJECT].EDIT_TOPIC, { ...instance, id: entityId }, user);
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/domain/stixCoreObject.js]: not sure to understand why it's necessary to notify here, since it's not done for stixCoreObjectImportPush. And why ABSTRACT_STIX_DOMAIN_OBJECT ?

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql [hunk 7]
```diff
@@ -4230,6 +4222,217 @@ input IndividualAddInput {
   file: Upload
 }
 
+################ Organizations
+enum OrganizationsOrdering {
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/config/schema/opencti.graphql]: all of this organizations part is in organization.graphql, maybe a merge issue ?

Comment (SarahBocognano) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql]: Yes It's a merge issue

File: opencti-platform/opencti-graphql/src/domain/stixCoreObject.js (hunk 6 above)
Comment (SarahBocognano) [reply opencti-platform/opencti-graphql/src/domain/stixCoreObject.js]: It's actually necessary to notify here because we delete the picture from the import component, so we need to notify the stix domain object (in our case threat actor individual) that a delete has been done in the stixcore import files

File: opencti-worker/src/config.yml.sample
Comment (SarahBocognano) [reply opencti-worker/src/config.yml.sample:1]: Because it wasn't supposed to be commit

**Comment by Archidoit** [comment]
What about making the boxes clickable ? (avoid to open the edition form) 
image

@SarahBocognano As discussed with Julien, we choose to make the boxes clickable inline disabled to keep the consistency of the platform. We plan to make editable "lines" in the future, so that way we can avoid forms sometimes, but for now it was asked to be disabled

**Comment by Archidoit** [comment]
Edit an image that has an order from Picture Management
Remove the 'order'
Update
An error occurs
image


That's strange because you can have images without order, but when an image has an order, you can't remove it.

**Comment by Archidoit** [comment]
When no order, we don't see the tooltip when the mouse is over it
image

**Comment by Archidoit** [comment]
What about enable the sort of the images by order in the carousel? (it's only a suggestion)
image

**Comment by Archidoit** [comment]
Error when going to an entity 'Data' tab.
Steps to reproduce:
- Go to a malware overview (or an other entity).
- Go to the 'Data' tab
- An error occurs

image

image

File: opencti-platform/opencti-front/src/private/components/common/cards/GenericAttackCard.tsx [hunk 8]
```diff
@@ -145,6 +160,7 @@ GenericAttackCardProps
       addBookmark(cardData.id, entityType);
     }
   };
+  const image = cardData.images?.edges ?? [];
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/cards/GenericAttackCard.tsx]: `images` in plural since it's an array.
Also, maybe it could be useful to have directly the avatarImage here, like this : 

```
const images = cardData.images?.edges ?? [];
const avatarImage = (images.length > 0 && images[0]) ? images[0].node : null;

```
And then you could directly use avatarImage, and only test if it's defined or not. What do you think ?

File: opencti-platform/opencti-front/src/private/components/common/cards/GenericAttackCard.tsx [hunk 9]
```diff
@@ -155,17 +171,17 @@ GenericAttackCardProps
...
-                alt={cardData.images[0].name}
+                src={getFileUri(image[0].node.id)}
+                alt={image[0].node.name}
               />
             ) : (
               <ItemIcon type={entityType} size="large" />
             )
             }
-            title={cardData.name}
+            title={renderThreatActorIndividual(cardData)}
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/cards/GenericAttackCard.tsx]: This component is generic and used for Malware, Campaign, Intrusion Sets, ... not only Threat Actors individuals. How does it work for them ?

File: opencti-platform/opencti-front/src/private/components/common/stix_domain_objects/StixDomainObjectBookmark.jsx [hunk 10]
```diff
@@ -162,7 +165,7 @@ class StixDomainObjectBookmarkComponent extends Component {
                 />
               </Avatar>
             )}
-            title={node.name}
+            title={renderThreatActorIndividual(node)}
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/stix_domain_objects/StixDomainObjectBookmark.jsx]: Same question here as for GenericAttackCard : is this method generic or specific for Threat Actors Individuals ? StixDomainObjectBookmark is a shared component.

File: opencti-platform/opencti-front/src/utils/String.jsx [hunk 11]
```diff
@@ -86,6 +89,34 @@ export const computeDuplicates = (fields, data) => R.groupWith(R.allPass(R.map(R
 
 export const capitalizeFirstLetter = (str) => str.charAt(0).toUpperCase() + str.slice(1);
 
+export const renderThreatActorIndividual = (threatActorIndividual) => {
+  if ((threatActorIndividual.locatedAtCountries?.edges ?? []).length > 0) {
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/utils/String.jsx]: Do we need more than one country ? If we need only one, maybe it could be handled by ThreatActorIndividual resolver to return the country ?

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql (hunk 7 above)
Comment (SouadHadjiat) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql]: it's still a duplicate no? you don't need to remove it ?

File: opencti-platform/opencti-graphql/src/database/file-storage.js [hunk 12]
```diff
@@ -304,15 +305,39 @@ export const upload = async (context, user, path, fileUpload, opts) => {
...
+              ...file.node.metaData,
+              order: elasticFile.order,
+              inCarousel: elasticFile.inCarousel,
+              description: elasticFile.description
+            }
+          }
+        };
+      }).sort((a, b) => {
+        const orderA = a.node.metaData?.order ?? Infinity;
+        const orderB = b.node.metaData?.order ?? Infinity;
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/database/file-storage.js]: What is Infinity ? can't we use 0 ?

**Comment by SouadHadjiat** [comment]
If I bookmark a malware, there is an error (frontend) and I can't load the malware list anymore : 

Capture d'écran 2023-08-25 205511

Happens for Campaigns too

=> now resolved

File: opencti-platform/opencti-front/src/private/components/common/stix_core_relationships/StixCoreRelationshipCreationFromEntity.jsx [hunk 13]
```diff
@@ -569,26 +569,26 @@ const StixCoreRelationshipCreationFromEntity = (props) => {
...
+                schema.schemaRelationsTypesMapping,
+              ),
+            );
+            return (
+                <>
+                  <div className={classes.header}>
+                    <IconButton
+                        aria-label="Close"
+                        className={classes.closeButton}
+                        onClick={() => handleClose()}
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/common/stix_core_relationships/StixCoreRelationshipCreationFromEntity.jsx]: I don't see anything that changed in this file, except identation (and not the right one)

**Comment by SouadHadjiat** [comment]
Error when opening a specific threat actor group (doesn't happen with all) : 

Capture d'écran 2023-08-25 222821

Investigating => Fixed

File: opencti-platform/opencti-front/src/private/components/threats/intrusion_sets/IntrusionSetCard.tsx [hunk 14]
```diff
@@ -31,6 +31,18 @@ const IntrusionSetCardFragment = graphql`
...
+            node {
+              id
+              name
+              metaData {
+                inCarousel
+                  description
+              }
+            }
+          }
+        }
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-front/src/private/components/threats/intrusion_sets/IntrusionSetCard.tsx:37]: I have a performance issue with this part on intrusion set list (and probably on Threat actors groups and individuals), since it's loaded for each element of the list : I guess it needs batching => on it

File: opencti-platform/opencti-graphql/src/resolvers/stixDomainObject.js [hunk 15]
```diff
@@ -96,7 +95,7 @@ const stixDomainObjectResolvers = {
...
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/resolvers/stixDomainObject.js:101]: I think this could be an issue, because we don't want to subscribe to all edit events for current user (to be confirmed @richard-julien)

File: opencti-platform/opencti-front/src/private/components/threats/intrusion_sets/IntrusionSetCard.tsx (hunk 14 above)
Comment (SouadHadjiat) [reply opencti-platform/opencti-front/src/private/components/threats/intrusion_sets/IntrusionSetCard.tsx:37]: resolved for cards by adding "avatar" field in graphql
